### PR TITLE
Basic GUI for Sensors

### PR DESCRIPTION
**Pull request by pfefferle**
#### Description
If you ask about a state of a (binary)sensor, it will display the value in a very basic form.

image

image

#### Type of PR
- [ ] Bugfix
- [X] Feature implementation
- [ ] Refactor of code (without functional changes)
- [ ] Documentation improvements
- [ ] Test improvements

#### Testing
Say "What is the value of {Binary Sensors Name}."

### CODE REVIEW

**Comment by pfefferle** [comment]
I will also add binary sensors when #82 is merged.

**Comment by stratus-ss** [comment]
@krisgesling can you take a look at this? I am not familiar with the GUI

**Comment by pfefferle** [comment]
Added binary_sensor support

image

image

**Comment by pfefferle** [comment]
It seems that the background image is persistant and stays also on other GUIs like settings or naptime.

@krisgesling is `skillBackgroundSource` the wrong method or do I miss a reset function?

File: ui/sensors.qml [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    property bool hasSensorDescription: sessionData.sensorDescription.length > 0 ? true : false
+
+    contentItem: ColumnLayout {
+        Label {
+            id: sensorName
+            Layout.fillWidth: true
+            font.pixelSize: 25
+            wrapMode: Text.Wrap
+            horizontalAlignment: Text.AlignHCenter
+            font.family: "Noto Sans"
```
Comment (krisgesling) [inline ui/sensors.qml]: We can start leaving these out unless you specifically want your Skill to use this font. 
This is now handled on a per device basis allowing different projects to create a more branded experience by setting their own default fonts and styling.

File: ui/sensors.qml [hunk 2]
```diff
@@ -0,0 +1,50 @@
+import QtQuick 2.4
+import QtQuick.Controls 2.2
+import QtQuick.Layouts 1.4
+import org.kde.kirigami 2.4 as Kirigami
+
+import Mycroft 1.0 as Mycroft
+
+Mycroft.Delegate {
+    id: homeAssistantRoot
+    skillBackgroundSource: Qt.resolvedUrl("wallpapers/home-assistant-default.png")
```
Comment (krisgesling) [inline ui/sensors.qml]: We've been moving to a new CardDelegate. This adds a very small gutter around the edge that will be used to produce effects indicating the status of the system. Eg it lights up blue when the wake word is detected. 

This is most important for devices that don't have LED's on top, but is also useful on the Mark II if you have it sitting on a shelf where you can't easily see the top.

So if you switch this to the following it should work nicely.
```qml
Mycroft.CardDelegate {
    id: homeAssistantRoot
    skillBackgroundColorOverlay: "transparent"
    cardBackgroundOverlayColor: "transparent"
    skillBackgroundSource: Qt.resolvedUrl("wallpapers/home-assistant-default.png")
```

File: __init__.py [hunk 3]
```diff
@@ -644,13 +663,29 @@ def _handle_set_thermostat(self, message):
...
+        )
+
+    def _display_sensor_dialog(self, name, value, description=""):
+        self.gui.clear()
+        self.gui["sensorName"] = name
+        self.gui["sensorValue"] = value
+        self.gui["sensorDescription"] = description
+        self.gui.show_page("sensors.qml")
+        time.sleep(15)
+        self.gui.release()
```
Comment (krisgesling) [inline __init__.py]: This can be simplified to:
```python
self.gui.show_page("sensors.qml", override_idle=15)
```

**Review by krisgesling — COMMENTED** [review]
Hey, exciting to see GUI elements coming in!

The image remaining across Skills is something we have to look at. Not a problem with your implementation. However I've made some little suggestions below that should work better.

I've also wondered if we should add an intent to display the HomeAssistants own GUI.
There's a handy `self.gui.show_url()` method.

**Comment by pfefferle** [comment]
> I've also wondered if we should add an intent to display the HomeAssistants own GUI. There's a handy self.gui.show_url() method.

@krisgesling I had a very similar idea, but wanted to start it as a stand alone skill, because I was not sure if it fits: https://github.com/pfefferle/skill-open-homeassistant

But now I will skip the idea and add it directly to this plugin.

File: ui/sensors.qml (hunk 1 above)
Comment (pfefferle) [reply ui/sensors.qml]: makes sense!

**Comment by pfefferle** [comment]
@krisgesling see https://github.com/MycroftAI/skill-homeassistant/pull/91

**Comment by pfefferle** [comment]
@krisgesling works perfect, thanks for the feedback!

**Comment by AIIX** [comment]
Can this be done natively in QML UI rather than using a webpage? For example, it would be nice to have something like this instead of opening a website 
devices-ui-1
devices-ui-2

**Comment by krisgesling** [comment]
Yeah to clarify, I didn't mean to use the web view instead of QML. Personally, I'd like us to end up with native QML views for all of the interactions, state reporting etc. And you could obviously toggle things from the GUI like turning lights on and off.

I think where the web view comes in is that a lot of HomeAssistant users have already spent a lot of time designing custom interfaces that suit their personal setup. So it would be good to let them use those same GUI's on their Mycroft devices if they have them and want to. These might be based on location groupings like everything in the kitchen, or to serve different functional purposes like alarm / cctv monitoring.

It will require some thinking around how we people can configure what is shown when, and I don't know how well they will display on a small screen, or if there's a way to show a lovelace view without the nav header which takes up a bunch of room especially if you have multiple tabs.

Also pinging @dschweppe as I know he's keen to input around the design side of things.

**Comment by pfefferle** [comment]
Should we open an issue for that? I think it is far bigger than the scope of that PR. I simply wanted to add a visual to the status check of the several sensors.

@krisgesling can you re-review my changes?

**Comment by AIIX** [comment]
> Should we open an issue for that? I think it is far bigger than the scope of that PR. I simply wanted to add a visual to the status check of the several sensors.

https://github.com/MycroftAI/skill-homeassistant/issues/93

**Comment by pfefferle** [comment]
"Open Home Assistant" was added with #91

File: __init__.py [hunk 4]
```diff
@@ -580,11 +580,24 @@ def _handle_sensor(self, message):
...
+                sensor_name,
+                attributes['current_temperature'],
+                sensor_state
+            )
         elif domain == "cover":
             self.speak_dialog(f'homeassistant.sensor.cover.{sensor_state}', data={
                 "dev_name": sensor_name})
+
+            sensor_states = self.translate_namedvalues(
+                'homeassistant.sensor.cover.state')
```
Comment (krisgesling) [inline __init__.py:594]: Could move these to `initialize()` so we only have to do it once

**Review by krisgesling — APPROVED** [review]
Looks good - I haven't run it on an install so would be great if someone could do that - but the new Github Actions should also catch any major issues. 

The GUI looks good on the few sizes I tested.

Small optional suggestion below

File: __init__.py (hunk 4 above)
Comment (krisgesling) [reply __init__.py:594]: Tests are passing so I'll merge this - and refactoring can happen in a follow up PR if there's interest in it.

**Pull request merged**